underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -106,14 +105,30 @@
   /** Whether the streaming upload is enabled. */
   private final boolean mStreamingUploadEnabled;
 
+//  private final Supplier<ObjectPermissions> mPermissions = this::getPermissionsInternal;
+
   /** The permissions associated with the bucket. Fetched once and assumed to be immutable. */
-  private final Supplier<ObjectPermissions> mPermissions = Suppliers
-      .memoize(new Supplier<ObjectPermissions>() {
-        @Override
-        public ObjectPermissions get() {
-          return getPermissionsInternal();
+  private final Supplier<ObjectPermissions> mPermissions = memoize(this::getPermissionsInternal);
+
+  /** Memoize implementation for java.util.function.supplier. */
+  private static <T> Supplier<T> memoize(Supplier<T> original) {
+    return new Supplier<T>() {
+      Supplier<T> mDelegate = this::firstTime;
+      boolean mInitialized;
+      public T get() {
+        return mDelegate.get();
+      }
+
+      private synchronized T firstTime() {
+        if (!mInitialized) {
+          T value = original.get();
+          mDelegate = () -> value;
```

Comment:
Don't we set `mDelegate` above as well?

Reply:
```mDelegate``` is declared as a Supplier, and here it works like a cache. It caches the value returned from the supplied get() method. So, after its first access to the ```Supplier<T> original```,  ```mDelegate``` is re-assigned with the returned value.